File: ts/src/hcanon.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */

// The HASH FORM (G6 phase 0, docs/capability-review/g6-distribution.md):
// exactly the unify-level canon with the additions that close its
// semantic gaps, so that two documents with the same hash form have the
// same meaning:
//
//   - a CLOSED map or list renders wrapped: close({...}), close([...])
//     (canon drops closedness);
//   - the type/hide MARKS render as their builtin wrappers: type(x),
//     hide(x) (canon drops marks).
//
// Both additions reuse existing parseable syntax, so the hash form
// remains valid Aontu source and round-trips:
// hcanon(unify(parse(hcanon(v)))) == hcanon(v) is a spec-suite property
// (test/spec/hcanon.tsv). User-facing canon is UNCHANGED — hcanon is a
// separate rendering.
//
// The marks PROPAGATE to every descendant at unification (walkMark), so
// a wrapper is emitted only where a mark STARTS: the walk carries the
// inherited marks down and a child whose mark the parent already
// carries renders bare. Without that, hide({a:1}) would render every
// leaf re-wrapped — still correct, never minimal, and not what the
// source said.
//
// canonHash (G6 phase 1) is the pin built on it:
//   "aon1-" + base64url( SHA-256( UTF-8( hcanon(v) ) ) )
// (unpadded base64url, RFC 4648 section 5). The "aon1-" scheme id
// exists so a future semantically-stronger normal form is an upgrade,
// not a breakage. This is a CANONICAL-TEXT hash, not a semantic
// equivalence class: the failure direction is the safe one — a false
// "changed" forces a needless re-review; a false "unchanged" would need
// two hash forms with the same bytes and different meanings, which is
// exactly the gap the close/mark wrappers exist to shut.

import { createHash } from 'node:crypto'

import type { Val } from './val/Val'

import { cmpCodePoint } from './keyorder'


type HMarks = {
  type: boolean
  hide: boolean
}


// One node's rendering, carrying the marks the ANCESTORS already
// wrapped. Bags and junctions recurse structurally (their canon getters
// render children through plain canon, which would drop a nested
// close); everything else — scalars, kinds, funcs, refs, constraints —
// delegates to its own canon, whose text is already in cross-port
// parity. The non-Val arm mirrors MapVal.canon's raw-peg fallback and
// is unreachable through an evaluated tree (direct-tested, ADR-002).
function render(v: any, inh: HMarks): string {
  if (true !== v?.isVal) {
    return String(v)
  }

  const mtype = true === v.mark?.type
  const mhide = true === v.mark?.hide
  const inner: HMarks = {
    type: inh.type || mtype,
    hide: inh.hide || mhide,
  }

  let s: string
  if (true === v.isMap) {
    const keys = Object.keys(v.peg).sort(cmpCodePoint)
    s = '{' +
      (v.spread.cj ? '&:' + render(v.spread.cj, inner) +
        (0 < keys.length ? ',' : '') : '') +
      keys.map((k) =>
        JSON.stringify(k) +
        (v.optionalKeys.includes(k) ? '?' : '') +
        ':' +
        render(v.peg[k], inner)).join(',') +
      '}'
    if (true === v.closed) {
      s = 'close(' + s + ')'
    }
  }
  else if (true === v.isList) {
    const keys = Object.keys(v.peg)
    s = '[' +
      (v.spread.cj ? '&:' + render(v.spread.cj, inner) +
        (0 < keys.length ? ',' : '') : '') +
      keys.map((k) => render(v.peg[k], inner)).join(',') +
      ']'
    if (true === v.closed) {
      s = 'close(' + s + ')'
    }
  }
  else if (true === v.isPref) {
    s = '*' + render(v.peg, inner)
  }
  else if (true === v.isConjunct || true === v.isDisjunct) {
    s = junctionText(v, true === v.isConjunct ? '&' : '|', inner)
  }
  else {
    s = v.canon
  }

  if (mtype && !inh.type) {
    s = 'type(' + s + ')'
  }
  if (mhide && !inh.hide) {
    s = 'hide(' + s + ')'
  }

  // The deprecation record rides outermost, as canonDeprecation
  // renders it (the wrappers are all reparseable calls, so order only
  // has to be FIXED, and this matches the canon the G3 rows pinned).
  const d = v.deprecation
  if (null != d) {
    const dkeys = Object.keys(d).sort()
    const rec = dkeys.map((k) =>
      JSON.stringify(k) + ':' + JSON.stringify(d[k])).join(',')
    s = 'deprecate(' + s + ('' === rec ? '' : ',{' + rec + '}') + ')'
  }

  return s
}


// The JunctionVal.canon parenthesisation rule, kept: a member that is
// itself a junction with more than one term keeps its parens so the
// text reparses with the same structure (`(1|2)&3`, not the
// differently-parsing `1|2&3`). Post-unification junctions are
// flattened by norm, so no SOURCE reaches the wrapping arm -- it is
// pinned by direct tests over constructed Vals in both ports, because
// a hash form that could render ambiguously would be a pin that
// silently agrees with a document it should not.
function junctionText(v: any, sym: string, inner: HMarks): string {
  return v.peg.map((m: any) =>
    true === m?.isJunction && 1 < m.peg.length
      ? '(' + render(m, inner) + ')'
      : render(m, inner))
    .join(sym)
}


// The hash form of an EVALUATED Val (unify first; parse-level canon
// parenthesisation differs between the ports and is excluded by
// construction — AGENTS.md).
export function hcanon(v: Val): string {
  return render(v, { type: false, hide: false })
}


// The canon-hash pin. Scoped to the module evaluated STANDALONE: its
// own include closure resolved and unified at its own root, before any
// consumer context — which is what makes the pin transitive (an edit
// two includes deep changes the unified root, hence the hash).
export function canonHash(v: Val): string {
  return 'aon1-' +
    createHash('sha256').update(hcanon(v), 'utf8').digest('base64url')
}

File: go/hcanon.go
/* Copyright (c) 2025 Richard Rodger, MIT License */

// The HASH FORM (G6 phase 0, the Go side of ts/src/hcanon.ts): exactly
// the unify-level canon with the additions that close its semantic
// gaps — a CLOSED map or list renders wrapped (close({...}),
// close([...])) and the type/hide MARKS render as their builtin
// wrappers (type(x), hide(x)). Both reuse parseable syntax, so the
// hash form remains valid Aontu source and round-trips
// (test/spec/hcanon.tsv). User-facing Canon() is UNCHANGED.
//
// The marks propagate to every descendant at unification (walkMark),
// so a wrapper is emitted only where a mark STARTS: the walk carries
// the inherited marks down and a child whose mark the parent already
// carries renders bare.
//
// CanonHash (G6 phase 1) is the pin built on it:
//
//	"aon1-" + base64url( SHA-256( UTF-8( Hcanon(v) ) ) )
//
// (unpadded base64url, RFC 4648 section 5). The "aon1-" scheme id
// exists so a future semantically-stronger normal form is an upgrade,
// not a breakage.

package aontu

import (
	"crypto/sha256"
	"encoding/base64"
	"sort"
	"strings"
)

type hcanonMarks struct {
	mtype bool
	mhide bool
}

// hcanonRender is one node's rendering, carrying the marks the
// ANCESTORS already wrapped. Bags and junctions recurse structurally
// (their Canon methods render children through plain Canon, which
// would drop a nested close); everything else — scalars, kinds, funcs,
// refs, constraints — delegates to its own Canon, whose text is
// already in cross-port parity.
func hcanonRender(v Val, inh hcanonMarks) string {
	if nil == v {
		return "nil"
	}

	mtype := v.markedType()
	mhide := v.markedHide()
	inner := hcanonMarks{
		mtype: inh.mtype || mtype,
		mhide: inh.mhide || mhide,
	}

	var s string
	switch b := v.(type) {
	case *MapVal:
		var out strings.Builder
		out.WriteByte('{')
		if b.spread != nil {
			out.WriteString("&:")
			out.WriteString(hcanonRender(b.spread, inner))
			if len(b.keys) > 0 {
				out.WriteByte(',')
			}
		}
		keys := append([]string(nil), b.keys...)
		sort.Strings(keys)
		for i, k := range keys {
			if i > 0 {
				out.WriteByte(',')
			}
			out.WriteString(jsonString(k))
			if b.isOptional(k) {
				out.WriteByte('?')
			}
			out.WriteByte(':')
			out.WriteString(hcanonRender(b.peg[k], inner))
		}
		out.WriteByte('}')
		s = out.String()
		if b.closed {
			s = "close(" + s + ")"
		}
	case *ListVal:
		var out strings.Builder
		out.WriteByte('[')
		if b.spread != nil {
			out.WriteString("&:")
			out.WriteString(hcanonRender(b.spread, inner))
			if len(b.peg) > 0 {
				out.WriteByte(',')
			}
		}
		for i, e := range b.peg {
			if i > 0 {
				out.WriteByte(',')
			}
			out.WriteString(hcanonRender(e, inner))
		}
		out.WriteByte(']')
		s = out.String()
		if b.closed {
			s = "close(" + s + ")"
		}
	case *PrefVal:
		s = "*" + hcanonRender(b.peg, inner)
	case *ConjunctVal:
		s = hcanonJunction(b.peg, "&", inner)
	case *DisjunctVal:
		s = hcanonJunction(b.peg, "|", inner)
	default:
		s = v.Canon()
	}

	if mtype && !inh.mtype {
		s = "type(" + s + ")"
	}
	if mhide && !inh.mhide {
		s = "hide(" + s + ")"
	}

	// The deprecation record rides outermost, as canonDeprecation
	// renders it (the wrappers are all reparseable calls, so order
	// only has to be FIXED, matching ts/src/hcanon.ts).
	if d := v.deprecRec(); nil != d {
		keys := make([]string, 0, len(d))
		for k := range d {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		rec := ""
		for i, k := range keys {
			if 0 < i {
				rec += ","
			}
			rec += jsonString(k) + ":" + jsonString(d[k])
		}
		if "" == rec {
			s = "deprecate(" + s + ")"
		} else {
			s = "deprecate(" + s + ",{" + rec + "})"
		}
	}

	return s
}

// hcanonJunction joins junction members, keeping junctChildCanon's
// parenthesisation rule exactly: a child that is itself a junction
// with more than one term keeps its parens so the text reparses with
// the same structure.
func hcanonJunction(members []Val, sym string, inner hcanonMarks) string {
	parts := make([]string, len(members))
	for i, m := range members {
		wrap := false
		switch t := m.(type) {
		case *ConjunctVal:
			wrap = len(t.peg) > 1
		case *DisjunctVal:
			wrap = len(t.peg) > 1
		}
		if wrap {
			parts[i] = "(" + hcanonRender(m, inner) + ")"
		} else {
			parts[i] = hcanonRender(m, inner)
		}
	}
	return strings.Join(parts, sym)
}

// Hcanon is the hash form of an EVALUATED Val (unify first;
// parse-level canon parenthesisation differs between the ports and is
// excluded by construction — AGENTS.md). Mirrors hcanon in
// ts/src/hcanon.ts.
func Hcanon(v Val) string {
	return hcanonRender(v, hcanonMarks{})
}

// CanonHash is the canon-hash pin. Scoped to the module evaluated
// STANDALONE: its own include closure resolved and unified at its own
// root, before any consumer context — which is what makes the pin
// transitive (an edit two includes deep changes the unified root,
// hence the hash). Mirrors canonHash in ts/src/hcanon.ts.
func CanonHash(v Val) string {
	sum := sha256.Sum256([]byte(Hcanon(v)))
	return "aon1-" + base64.RawURLEncoding.EncodeToString(sum[:])
}
